mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -212,12 +230,140 @@ class LinearRegressionModel private[ml] (
+    case Some(summ) => summ
+    case None =>
+      throw new SparkException(
+        "No training summary available for this LinearRegressionModel",
+        new NullPointerException())
+  }
+
+  private[regression] def setSummary(summary: LinearRegressionTrainingSummary): this.type = {
+    this.trainingSummary = Some(summary)
+    this
+  }
+
+  def hasSummary: Boolean = trainingSummary.isDefined
+
+
+  /**
+   * Evaluates the model on a testset.
+   * @param dataset Test dataset to evaluate model on.
+   */
+  def evaluate(dataset: DataFrame): LinearRegressionSummary = {
+    val t = udf { features: Vector => predict(features) }
+    val predictionAndObservations = dataset
+      .select(col($(labelCol)), t(col($(featuresCol))).as($(predictionCol)))
+
+    new LinearRegressionSummary(predictionAndObservations, $(predictionCol), $(labelCol))
+  }
+
   override protected def predict(features: Vector): Double = {
     dot(features, weights) + intercept
   }
 
   override def copy(extra: ParamMap): LinearRegressionModel = {
-    copyValues(new LinearRegressionModel(uid, weights, intercept), extra)
+    val newModel = new LinearRegressionModel(uid, weights, intercept)
+    if (trainingSummary.isDefined) newModel.setSummary(trainingSummary.get)
+    copyValues(newModel, extra)
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Linear regression training results.
+ * @param predictions predictions outputted by the model's `transform` method.
+ * @param objectiveTrace objective function (scaled loss + regularization) at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingSummary private[ml] (
+    predictions: DataFrame,
+    predictionCol: String,
+    labelCol: String,
+    val objectiveTrace: DataFrame)
+  extends LinearRegressionSummary(predictions, predictionCol, labelCol) {
+
+  def this(
+      predictions: DataFrame,
+      predictionCol: String,
+      labelCol: String,
+      objectiveTrace: Array[Double]) = {
+
+    this(
+      predictions,
+      predictionCol,
+      labelCol,
+      predictions.sqlContext
+        .createDataFrame(objectiveTrace.zipWithIndex.map(_.swap).toSeq)
+        .toDF("iteration", "objective")
+    )
+  }
+
+  /** Number of training iterations until termination */
+  val totalIterations = objectiveTrace.count()
+
+}
+
+/**
+ * :: Experimental ::
+ * Linear regression results evaluated on a dataset.
+ * @param predictions predictions outputted by the model's `transform` method.
+ */
+@Experimental
+class LinearRegressionSummary private[ml] (
+    @transient val predictions: DataFrame,
+    val predictionCol: String,
+    val labelCol: String) extends Serializable {
+
+  private val metrics = new RegressionMetrics(predictions
+    .select(predictionCol, labelCol)
+    .map { case Row(pred: Double, label: Double) => (pred, label) }
+  )
+
+  // Force evaluation of lazy RegressionMetrics.summary for proper serialization
+  metrics.explainedVariance
```

Comment:
I have a hard time thinking of why a worker needs to have the objective trace; that's the sort of thing people would just want to print from the driver.  I'm sure that, at some point, someone will need workers to access some stats, but those can be extracted on the driver & then passed to the workers.